components/script/dom/workers/serviceworkercontainer.rs

```diff
@@ -181,6 +184,164 @@ impl ServiceWorkerContainerMethods<crate::DomTypeHolder> for ServiceWorkerContai
         // A: Step 7
         promise
     }
+
+    /// <https://w3c.github.io/ServiceWorker/#navigator-service-worker-getRegistration>
+    fn GetRegistration(&self, client_url: USVString) -> Rc<Promise> {
+        let global = self.global();
+        // Step 1. Let client be this’s service worker client.
+        // self.client is the client.
+        // Step 2. Let storage key be the result of running obtain a storage key given client.
+        let storage_key = match self.client.obtain_storage_key() {
+            Ok(key) => key,
+            Err(()) => {
+                let promise = Promise::new(&global, CanGc::note());
+                promise.reject_error(Error::InvalidAccess, CanGc::note());
+                return promise;
+            },
+        };
+        // Step 3. Let clientURL be the result of parsing clientURL with this’s relevant settings object’s API base URL.
+        let mut client_url = match global.api_base_url().join(client_url.0.as_str()) {
+            Ok(url) => url,
+            Err(_) => {
+                // Step 4. If clientURL is failure, return a promise rejected with a TypeError.
+                let promise = Promise::new(&global, CanGc::note());
+                promise.reject_error(Error::Type("Invalid URL".to_string()), CanGc::note());
+                return promise;
+            },
+        };
+        // Step 5. Set clientURL’s fragment to null.
+        client_url.set_fragment(None);
+        // Step 6. If the origin of clientURL is not client’s origin, return a promise rejected with a "SecurityError" DOMException.
+        if client_url.origin() != self.client.creation_url().origin() {
+            let promise = Promise::new(&global, CanGc::note());
+            promise.reject_error(Error::Security, CanGc::note());
+            return promise;
+        }
+        // Step 7. Let promise be a new promise.
+        let promise = Promise::new(&global, CanGc::note());
+        // Step 8. Run the following substeps in parallel:
+        let script_to_constellation_chan = global.script_to_constellation_chan();
+        let (sender, receiver) = ipc::channel().unwrap();
+        let task_source = global
+            .task_manager()
+            .dom_manipulation_task_source()
+            .to_sendable();
+        // Step 8.1. Let registration be the result of running Match Service Worker Registration given storage key and clientURL.
+        let message = ScriptToConstellationMessage::MatchRegistration {
+            storage_key,
+            url: client_url,
+            sender: sender.clone(),
+        };
+        let _ = script_to_constellation_chan.send(message);
+        ROUTER.add_typed_route(
```

Comment:
We should not setup routes per operation, but rather setup some kind of persistent route for all communications between a client and the service worker. Not to be addressed for now.